### PR TITLE
On connect dialog, in the directory list, remove 'Custom' and add all custom central servers

### PR DESCRIPTION
**Pull request by jp8**
On the connect dialog, in the directory (custom central server) list,
- remove Custom from list
- add the user's custom central servers to the list

Addresses discussion #1869
Is an alternative to pull request #1884 

It is a change to the UI.  If the pull request is accepted, I will make a corresponding pull request to the website documentation.

Working implementation

It's a UI change so testing and feedback are appreciated.  It is one alternative UI among several that are under discussion in #1869 and #1884 

## Checklist

- [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
- [x] I tested my code and it does what I want
- [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency) 
- [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
- [x] I've filled all the content above

 ## Last steps:
 
- [ ]  Read through code and test it (should be done already)
 - [x]  Add @jp8 to contributor page
 - [ ]  Add Change log entry
 - [x]  Get two approvals

### CODE REVIEW

**Comment by ann0see** [comment]
Hi @jp8 thank you for your contribution. The style doesn't seem to be right. Could you please have a look at clang format?

Also would it be possible to have screenshots?

**Comment by jp8** [comment]
Thanks @ann0see 

clang format is in progress.

Screenshots copied from #1884 

Screenshot_2021-06-26_08-18-23
Screenshot_2021-06-26_08-20-31

**Comment by ann0see** [comment]
Ok. Wasn't up to date with the other PR. Once the artifacts are built - and I find time - I'd be happy to test your proposal. From a first glance it looks interesting ;-)

**Comment by ann0see** [comment]
BTW, you can run `clang-format -i /path/to/files` locally to fix the code style.

**Comment by jp8** [comment]
Thanks - I did try that approach several weeks ago but got errors, likely due to an old clang-format in Debian 10.

**Comment by jp8** [comment]
I tried to write it in a way that would remain compatible with all the other proposals that are still under discussion in #1884

**Comment by ann0see** [comment]
Yeah. Not a big deal since the check now seems to be ok.

After having compiled it I'd say I'm quite happy with this change! 
Sure, it doesn't follow the original request from the discussion, however, I think moving directory management from settings to the connect dialog might become a religious discussion ;-).

What about:

1. Reviewing and merging this PR
2. Adding an "edit" button which opens settings in the connect dialog?

File: src/connectdlg.cpp
Comment (ann0see) [inline src/connectdlg.cpp:71]: This comment is no longer true? --> move it to method

File: src/connectdlg.cpp
Comment (ann0see) [inline src/connectdlg.cpp:901]: Move the comment here (at least that would be my preference)?

File: src/connectdlg.cpp
Comment (ann0see) [inline src/connectdlg.cpp:910]:
```suggestion

    for ( int iLEIdx = 0; iLEIdx < MAX_NUM_SERVER_ADDR_ITEMS; iLEIdx++ )
```
Just a personal preference

**Review by ann0see — COMMENTED** [review]
Quite small but effective ;-)

**Comment by jp8** [comment]
>     2\. Adding an "edit" button which opens settings in the connect dialog?

I do think an edit button or link that opens the settings dialog to the correct page is the likely best way to get to the management interface.

But there might be a lot of discussion and ideas around that, and I don't think this small pull request depends on it.  Could we save it as a separate topic ?

File: src/connectdlg.cpp
Comment (pljones) [inline src/connectdlg.cpp:212]: This needs a bit more explanation now.

There are two fields being added together and it's not clear how the two relate.  I can see an address type being turned into an integer - fine, make sense as an index.  I can see a "custom directory index" - but no explanation what that means, so I guess it's an index into a list of custom directory servers.  My initial thought is that both could be non-zero.  Or both zero and be valid.

So I then think I must be missing something.  It's that something that needs explaining here.

File: src/connectdlg.cpp
Comment (pljones) [inline src/connectdlg.cpp:220]: Again, the code no longer seems to agree with this comment, so something needs changing.

File: src/connectdlg.cpp
Comment (pljones) [inline src/connectdlg.cpp:246]: And again, just a little more comment here to say how this works - it's pretty clear but writing it out in clear English to say what it's _intended_ to do lets people confirm that's what it's doing :).

File: src/connectdlg.cpp
Comment (pljones) [inline src/connectdlg.cpp:539]: This should only call RequestServerList if the custon directory being changed is the selected one.

File: src/connectdlg.cpp
Comment (pljones) [inline src/connectdlg.cpp:529]: This block needs explaining.

File: src/connectdlg.cpp
Comment (pljones) [inline src/connectdlg.cpp:911]: What's "LE"?  If it's "list entry" that doesn't add anything other than confusion.  Otherwise, I'd rather a longer name.

File: src/settings.h
Comment (pljones) [inline src/settings.h:160]: ?

File: src/connectdlg.cpp
Comment (pljones) [inline src/connectdlg.cpp:214]: "... and is thus the offset into cbxCentServAddrType after the last non-custom directory server"

File: src/connectdlg.cpp
Comment (jp8) [reply src/connectdlg.cpp:542]: i agree, but this call to RequestServerList() is from the existing functionaliry.  Volker put it in so that if you change the custom central server on the settings dialog, it updates immediately on the connect dialog.

I am just thinking to leave it alone until we have further discussions on how the custom server management should work.

Comment (pljones) [reply src/connectdlg.cpp:542]: Existing was
```
    // only update list if the custom server list is selected
    if ( pSettings->eCentralServerAddressType == AT_CUSTOM )
    {
        RequestServerList();
    }
```
So no, only if the custom server is the selected _one_ and it's changed (i.e. in settings a different custom server is selected).

We no longer _care_ if the _selected_ custom server _in settings_ changes.  That shouldn't cause a change in the connect dialogue.

We care if the definition of the server selected in _here_ is changed - but that's new and I don't think it's part of this.

File: src/connectdlg.cpp
Comment (jp8) [reply src/connectdlg.cpp:214]: hmm the combo box name doesn't really reflect its true function anymore.

File: src/connectdlg.cpp
Comment (jp8) [reply src/connectdlg.cpp:543]: I don't think so, because the settings dialog puts any modified or selected custom central server at position 0.

Comment (jp8) [reply src/connectdlg.cpp:543]: ah i see

File: src/connectdlg.cpp
Comment (pljones) [reply src/connectdlg.cpp:915]: I see where you got the "LE" from now...  I still don't like it...

Comment (jp8) [reply src/connectdlg.cpp:915]: It's a verbatim cut-and-paste from the settings dialog.  I don't know what it stands for.

File: src/connectdlg.cpp
Comment (jp8) [reply src/connectdlg.cpp]: I have put back the missing check for AT_CUSTOM and now the functionality is comparable to how it was before - the server list is now not requested if one of the built-in central servers was selected.

Your proposed solution is definitely better - I will try to make it work.

File: src/connectdlg.cpp
Comment (pljones) [reply src/connectdlg.cpp]: `cbxDirectoryServer` would do.  Feel free :)

File: src/settings.h
Comment (pljones) [inline src/settings.h:160]: I guess my concern is that this appears in `src/settings.h` but not `src/settings.cpp` -- I'd expect at least to see something initialising it - the constructor for `CClientSettings`, for example, like `eCentralServerAddressType ( AT_DEFAULT ),`.  (I don't like letting the compiler decide on values.  Yes, it'll _probably_ put a zero in... if it bothers to generate code to do that.)

**Review by dcorson-ticino-com — COMMENTED** [review]
Sorry I can't really comment, this week on vacation with only a cellphone.

File: src/settings.h
Comment (jp8) [reply src/settings.h:160]: Yes you are right.

Comment (jp8) [reply src/settings.h:160]: So funny the file is named .ini

File: src/settings.cpp
Comment (ann0see) [inline src/settings.cpp:451]: No space needed after iValue

Comment (jp8) [reply src/settings.cpp]: yes local clang-format was invented for people like me

File: src/connectdlg.cpp
Comment (pljones) [reply src/connectdlg.cpp]: You can't avoid re-order if you delete an entry - and if it was the currently selected entry, you're definitely on a different directory server.  So something _will_ be needed here to track what's happening in the client settings dialogue.

Comment (jp8) [reply src/connectdlg.cpp]: My view is that if the settings dialog allows re-ordering, then the settings dialog should calculate and return a corrected index.

Right now, if a user changes, on the settings dialog, the list of custom central servers, the connect dialog selects the first custom central server in the list.

I could:
1. leave it as it is
2. reset to the Default built-in central server after any management of the list of custom central servers
3. start modifying the settings dialog to send an index
4. handle it in connect dialog by comparing strings

None is ideal.  But users don't often modify their list of custom central servers.  If they do modify the list, I think option 2 seems like reasonable behaviour.  To me, option 3 seems premature and option 4 seems like it could be a step in the wrong direction.

Comment (jp8) [reply src/connectdlg.cpp]: the latest commit implements option 2 above, just so that anyone who is testing can get a feel for what I personally think is the best approach

File: src/connectdlg.cpp
Comment (jp8) [reply src/connectdlg.cpp:919]: iCustomDirectoryVectorIterator ?

**Comment by pljones** [comment]
By the way, feel free to
```
git checkout master
git pull upstream master; git pull upstream master --tags --force
git push; git push --tags --force
git checkout -
```
to refresh your local git and then
```
git rebase -i master
```
and squash history (use "fix" if you want).  I don't think any of us review on a commit by commit basis and the extra commit history will need cleaning up before it's merged.

File: src/connectdlg.cpp
Comment (pljones) [inline src/connectdlg.cpp:533]: Can you
1. Add a "TODO" for only updating the list of the actual directory server selected changed (i.e. remember where we were and see if we can find it in the new list).
2. Drop the other comments here.

I'm okay with this approach as a first cut, but I'd like to see an issue raised - and linked from the TODO - covering point 1.

File: src/connectdlg.cpp
Comment (pljones) [reply src/connectdlg.cpp:919]: :) `i` would do, in my view.  There are places with `iIdx` (an integer index...) but I don't even like that...

File: src/settings.cpp
Comment (pljones) [inline src/settings.cpp:455]: "Note: this is the correct value if `eCentralServerAddressType` is currently `AT_CUSTOM`."

(And if it isn't, set `eCentralServerAddressType` is currently `AT_DEFAULT`...)

File: src/connectdlg.cpp
Comment (jp8) [reply src/connectdlg.cpp:919]: >  I'd rather a longer name.

yeah it's either i, or iCustomDirectoryVectorIteratorWithNameSuggestedByPljones

**Comment by hoffie** [comment]
> On the connect dialog, in the directory (custom central server) list,
> 
>   * remove Custom from list
> 
>   * add the user's custom central servers to the list

I like this and I think this is a UI improvement, obviously mainly for users with custom directory servers (but without disturbing those who don't use them).

> Is an alternative to pull request #1884

I don't fully agree. #1884's intention -- as far as I understand -- was to make (initial) usage of custom directory servers easier. I still agree with the arguments there. I'm wondering if this UI improvement could be combined with what @dcorson-ticino-com and @gilgongo proposed in https://github.com/jamulussoftware/jamulus/pull/1884#issuecomment-864391060 -- i.e. showing a "Enter custom address" entry which would then disable the custom address field. Or at least adding a "Open settings to add custom servers" entry. I'm really only adding this here, as I don't consider this PR to be a full alternative to #1884 and I think this discussion can be postponed.

Haven't looked at the code yet, I'm just judging by the description and screenshots.

File: src/connectdlg.cpp
Comment (jp8) [reply src/connectdlg.cpp:533]: Yes I agree with you about adding a TODO but I think a discussion is needed around the broad topic of user interface for finding and connecting to servers, including:
- directory management
- server connection history
- server connection favourites/bookmarks
- ad-hoc server connection
- parental controls

Maybe there is already an appropriate discussion open ?  #1774 seems relevant.

I would like to propose a mock-up to address that part of the UI in a way that accommodates all types of users.

**Review by ann0see — CHANGES_REQUESTED** [review]
Please use pljones suggestion to rebase your work: https://github.com/jamulussoftware/jamulus/pull/1894#issuecomment-870714146

Doing it with a merge makes stuff complicated

**Comment by gilgongo** [comment]
@hoffie 

>  I still agree with the arguments there.

Sorry to be a pain but in the interests of clarity, can you detail why you agree, against the points made here? https://github.com/jamulussoftware/jamulus/pull/1884#issuecomment-867127479

**Comment by jp8** [comment]
> Please use pljones suggestion to rebase your work: [#1894 (comment)](https://github.com/jamulussoftware/jamulus/pull/1894#issuecomment-870714146)
> 
> Doing it with a merge makes stuff complicated

I am really sorry to be so new to git.  I did follow the steps from pljones but on my next attempt to push, it left me with:

hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart.

...and asked me to pull from my feature branch to local and then merge conflicts.  I am guessing that after following the steps, I should have pulled from my feature branch, before doing any further work.

**Comment by ann0see** [comment]
No. That’s what `git push --force` is used for ;-). However you might want to give https://git-scm.com/docs/git-rebase a short read.

**Comment by jp8** [comment]
> I‘d suggest to undo the merge, and then force push everything.

OK I will figure out how to do that.  If I understand right, the goal is to take the 20 commits that are now on the feature branch in my fork, and condense them into just one or a few commits... is it right ?

**Comment by ann0see** [comment]
No. You‘d git checkout this commit: https://github.com/jamulussoftware/jamulus/pull/1894/commits/79559e5fb1de9967fe4ed0b61ddeb8bda6d9bc6b 

git push HEAD to your feature branch and then use git rebase -i to bring the commits together.

**Comment by jp8** [comment]
> git push HEAD to your feature branch (https://stackoverflow.com/a/40929378) and then use git rebase -i to bring the commits together.

I think I've done all this.  I condensed it to 7 commits.  Is it ok ?

**Review by ann0see — COMMENTED** [review]
From a first glance this looks much better now ;-)

**Comment by jp8** [comment]
Is there anything more I need to do, in order to get this merged ?

**Comment by ann0see** [comment]
There are still some unresolved review comments above.

File: src/settings.cpp
Comment (jp8) [reply src/settings.cpp:455]: I think I understand your point here - a manual edit to the config file could leave it with an invalid combination of values.  I think my latest (final?) commit takes care of it.

**Comment by jp8** [comment]
ok all done i think.

**Comment by ann0see** [comment]
Last steps:
- [ ] Read through code and test it (should be done alread)
- [ ] Add @jp8 to contributor page
- [ ] Add Change log entry
- [ ] Get two approvals

Anything else? Could you please copy and paste this check list in your first post above?

**Comment by jp8** [comment]
>     * [ ]  Add @jp8 to contributor page

I was already on this list, from contributions I gave directly to Volker in the past.

**Review by pljones — APPROVED** [review]
I'm happy with it now.  (Subject to the TODO item getting done ASAP.)

**Review by ann0see — APPROVED** [review]
Looks OK to me. I‘ve tested an early version and the code looks ok.

**Comment by ann0see** [comment]
This should be squash&merged.

**Comment by pljones** [comment]
I couldn't get the github UI to do what I wanted, so I did it locally, hence I'm going to close this.
https://github.com/jamulussoftware/jamulus/commit/62a9c3831f88cdf5180a8f608fba3947504818d5

**Comment by ann0see** [comment]
There’s an arrow next to the merge button with which you can change it to squash and merge?

**Comment by pljones** [comment]
> There’s an arrow next to the merge button with which you can change it to squash and merge?

I don't think that lets you adjust the commit message and it's as if the branch never existed and a single commit to master happened directly.  At least I think that's what happened when I first tried it.  I wanted to show a merge from a branch with a single commit from jp8 with a sane commit message.

**Comment by jp8** [comment]
So at the end I should have sqashed again, into a single commit ?  (just so I know for next time...)

**Comment by pljones** [comment]
Yeah, it can sometimes help to keep the history (it recently let me split a change into to two PRs) but generally, once final, one commit is preferred.

**Pull request closed without merging**
